src/bgw/scheduler.c

```diff
@@ -325,19 +330,110 @@ scheduled_bgw_job_start(ScheduledBgwJob *sjob, register_background_worker_callba
 	}
 }
 
-static List *
-initialize_scheduled_jobs_list(MemoryContext mctx)
+static void
+terminate_and_cleanup_job(ScheduledBgwJob *sjob)
 {
-	List	   *jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
-	ListCell   *lc;
+	if (sjob->handle != NULL)
+	{
+		TerminateBackgroundWorker(sjob->handle);
+		WaitForBackgroundWorkerShutdown(sjob->handle);
+		pfree(sjob->handle);
+	}
+	sjob->may_need_mark_end = false;
+	worker_state_cleanup(sjob);
+}
 
-	foreach(lc, jobs)
+/*
+ *  Update the given job list with whatever is in the bgw_job table. For overlapping jobs,
+ *  copy over any existing scheduler info from the given jobs list.
+ *  Assume that both lists are ordered by job ID.
+ *  Note that this function call will destroy cur_jobs_list and return a new list.
+ */
+List *
+update_scheduled_jobs_list(List *cur_jobs_list, MemoryContext mctx)
+{
+	List	   *new_jobs = bgw_job_get_all(sizeof(ScheduledBgwJob), mctx);
+	ListCell   *new_ptr = list_head(new_jobs);
+	ListCell   *cur_ptr = list_head(cur_jobs_list);
+
+	while (cur_ptr != NULL && new_ptr != NULL)
 	{
-		ScheduledBgwJob *sjob = lfirst(lc);
+		ScheduledBgwJob *new_sjob = lfirst(new_ptr);
+		ScheduledBgwJob *cur_sjob = lfirst(cur_ptr);
+
+		if (cur_sjob->job.fd.id < new_sjob->job.fd.id)
+		{
+			/*
+			 * We don't need cur_sjob anymore. Make sure to clean up the job
+			 * state. Then keep advancing cur pointer until we catch up.
+			 */
+			terminate_and_cleanup_job(cur_sjob);
+
```

Comment:
should we pfree cur_sjob itself here. (also below where we also do cleanup)?

Reply:
No, we should not pfree here or in `worker_state_cleanup`, since the list of sjobs persists for the entire duration of the interval in between when the scheduler reloads jobs. The sjob will get freed in `list_free` below, whenever we destroy the entire list of jobs.

Reply:
then I think you need to use list_free_deep?

Reply:
Yes, good point. I misread the source and thought `list_free` also deletes `lfirst`. Fixed.